game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveSubsRetreat.java

```diff
@@ -0,0 +1,93 @@
+package games.strategy.triplea.delegate.battle.steps.retreat;
+
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.SUBS_SUBMERGE;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.SUBS_WITHDRAW;
+import static games.strategy.triplea.delegate.battle.steps.BattleStep.Order.SUB_OFFENSIVE_RETREAT_AFTER_BATTLE;
+import static games.strategy.triplea.delegate.battle.steps.BattleStep.Order.SUB_OFFENSIVE_RETREAT_BEFORE_BATTLE;
+
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle.RetreatType;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.List;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class OffensiveSubsRetreat implements BattleStep {
+
+  private static final long serialVersionUID = -244024585102561887L;
+
+  protected final BattleState battleState;
+
+  protected final BattleActions battleActions;
+
+  @Override
+  public List<String> getNames() {
+    if (!isEvaderPresent() || !isRetreatPossible()) {
+      return List.of();
+    }
+
+    if (getOrder() == SUB_OFFENSIVE_RETREAT_BEFORE_BATTLE && isDestroyerPresent()) {
```

Comment:
I'm not sure about keeping this logic.  It only works as long as nothing can kill the destroyer before the sub has a chance to retreat.  Since the sub can retreat at the beginning of the battle, the only thing that fires before its retreat is AA guns.  And if a map has an AA gun that can target destroyers, then an error could occur.

Would you expect AA guns that can target destroyers?  I could either remove this logic or change it to go through all of the units, figure out if there are any AA guns and determine if they can target all of the destroyers in the defenders side.  I lean towards removing the logic.  It'll mean that the withdraw/submerge step will show up more often in the UI then it did previously.

Reply:
It seems a bit questionable either way. Not sure I have a strong opinion to remove it. I'd opt personally to leave as-is for now. If changed, would be good to have it be a change on its own and a complete analysis of what we're updating (how often we see the display when we don't need to).

I think the real fix is to make the concepts more generic and check for those. 

IE: `firstStrikeUnit` and `firstStrikeCounterUnit`. You then have a 'first strike' label present and it's neutralized if you have a counter unit. We also need to keep in mind that the rulesets can vary as well and that we may need to make whether that happens or not completely configurable.

Reply:
Even if the concepts become more generic, the problem still exists.  The main reasons are:
* A battle has up to 3 places where casualties are removed and can't fire back.  This is after the AA battle, the sub battle, and bombardment.
* The collection of steps to show to the user happen at the very beginning of a round before any steps actually occur.

So if there is a `firstStrikeUnit` and a `firstStrikeCounterUnit`, you would normally think that the `firstStrikeWithdrawStep` would not happen since there is a counter unit.  But if the AA part happens before the `firstStrikeWithdrawStep` and the `firstStrikeCounterUnit` is killed, it is removed from the battle and no longer prevents the `firstStrikeWithdrawStep`.  So you would actually need to show the `firstStrikeWithdrawStep`.

But I'll leave the check here since it was how it was originally.

Reply:
If a step is valid at the beginning of combat, and then is removed, that could be fine. Perhaps nicer to visualize that it no longer applied. Overall, AFAIK, the number one requested feature for combat phases is the ability to "walk back in time" by clicking the phases to see what happened during that phase - very similar to how the history mode works.

FWIW, keeping as-is for now seems prudent, keeping this just to a refactor update.  :+1: